### PR TITLE
Pair as product in Idris

### PR DESCRIPTION
**Pull request by Jake-Gillberg**
Added a concrete implementation of Pair as Product in Idris, (similar to Either as CoProduct).

EitherAsCoProduct.lidr was missing `access public export` and `default total` directives, so I added those as well.

### CODE REVIEW

**Comment by Jake-Gillberg** [comment]
also, still quite new to idris and this library, so please let me know if I could make anything cleaner or more idiomatic!

**Comment by clayrat** [comment]
I think sticking to `camelCase` everywhere would make it a tiny bit more idiomatic ;)

**Comment by Jake-Gillberg** [comment]
> I think sticking to `camelCase` everywhere would make it a tiny bit more idiomatic ;)

looking at the names of things in Coq got my head twisted :)

File: src/Idris/PairAsProduct.lidr [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+>   rewrite hsnd in
+>   rewrite (surjectivePairing p2) in
+>   Refl
+>
+> equalAssoc : a = b -> b = a
+> equalAssoc Refl = Refl
+>
+> extensionalTypeMorphismEq : MkExtensionalTypeMorphism x = MkExtensionalTypeMorphism y
+>                          -> x = y
+> extensionalTypeMorphismEq Refl = Refl
```
Comment (Jake-Gillberg) [inline src/Idris/PairAsProduct.lidr]: Does this belong in TypesAsCategoryExtensional.lidr?

File: src/Idris/PairAsProduct.lidr [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+>                      -> Prelude.Basics.snd p1 = Prelude.Basics.snd p2
+>                      -> p1 = p2
+> injectiveProjections (a, b) p2 hfst hsnd =
+>   rewrite hfst in
+>   rewrite hsnd in
+>   rewrite (surjectivePairing p2) in
+>   Refl
+>
+> equalAssoc : a = b -> b = a
+> equalAssoc Refl = Refl
```
Comment (Jake-Gillberg) [inline src/Idris/PairAsProduct.lidr]: Do these exist in a standard library somewhere?

Comment (marcosh) [reply src/Idris/PairAsProduct.lidr]: if you search in the repl using `:s a = b -> b = a` you'll see that this is exactly `sym`

File: src/Idris/PairAsProduct.lidr (hunk 1 above)
Comment (marcosh) [reply src/Idris/PairAsProduct.lidr]: yes, makes sense, I moved it there

**Comment by marcosh** [comment]
Thanks @Jake-Gillberg, I merged it manually after removing some useless arguments (also from `EitherAsCoProduct`)

**Pull request closed without merging**